Add typescript support

yardnsm opened the pull request:
This PR attempts to add support for typescript type declarations.

viliusl commented:
Thanks @yardnsm for a PR, but. I would really appreciate if you could also add tests for typings. It might seem strange, but it's another asset that contributors will have to take into account, I will also have to keep in mind when doing changes etc. Having tests would allow me to easily assess if PR is correct or not - now it's only assumptions.

You could take a look at set-up and approach I started facoring at https://github.com/wix-platform/wix-node-platform/blob/master/security/wix-server-to-server-signer/package.json#L12 and https://github.com/wix-platform/wix-node-platform/blob/master/security/wix-server-to-server-signer/test/typings.spec.ts

Idea is to not test behavior, but more of a form - what is exposed, what is returned, do types match, verify permutations of optional params - if runtime matches dts. Also I see that .d.ts is only partial where https://github.com/wix/wix-eventually/blob/master/index.js#L27 is not covered:/

wdyt?

yardnsm commented:
Just added tests for the declarations, similar to the example you've sent. There is a problem with `eventually.with`, since `with` is a reserved word and TypeScript expects a `with` statement to exist.

I've opened an issue in TypeScript regarding this issue: https://github.com/Microsoft/TypeScript/issues/21723

viliusl commented on index.d.ts:
```diff
@@ -0,0 +1,10 @@
+
+declare namespace WixEventually {
+  export interface Opts {
+    timeout: number;
```
both options are optional

viliusl commented on test/typings.spec.ts:
```diff
@@ -0,0 +1,17 @@
+
+describe('wix-eventually types', () => {
+  it('exports a function', () => {
+    expect(eventually).to.not.be.undefined;
```
you can drop this - afaik second one asserts undefined

yardnsm commented:
I updated the tests and added support for the `with` function. I didn't understand your intention on behavioral testing the options, can you please explain it again? :)

viliusl commented:
Thanks alot!
